src/Common/OvercommitTracker.cpp

```diff
@@ -115,24 +115,38 @@ void OvercommitTracker::tryContinueQueryExecutionAfterFree(Int64 amount)
     if (OvercommitTrackerBlockerInThread::isBlocked())
         return;
 
-    std::lock_guard guard(overcommit_m);
+    std::shared_lock read_lock(overcommit_m);
     if (cancellation_state != QueryCancellationState::NONE)
     {
-        freed_memory += amount;
-        if (freed_memory >= required_memory)
-            releaseThreads();
+        read_lock.unlock();
+        {
+            std::lock_guard lk(overcommit_m);
+            if (cancellation_state != QueryCancellationState::NONE)
+            {
+                freed_memory += amount;
+                if (freed_memory >= required_memory)
+                    releaseThreads();
+            }
+        }
     }
```

Comment:
Maybe early return to reduce code complexity?
```suggestion
    std::shared_lock read_lock(overcommit_m);
    if (cancellation_state = QueryCancellationState::NONE)
        return;
    read_lock.unlock();
    
    std::lock_guard lk(overcommit_m);
    if (cancellation_state != QueryCancellationState::NONE)
    {
        freed_memory += amount;
        if (freed_memory >= required_memory)
            releaseThreads();
    }
```

Reply:
Thanks for the suggestion and It looks good. I just wonder do we need to unlock the `read_lock` first before acquire the `write_lock`?

Reply:
I think `read_lock` will be destroyed as we leave the scope right before `lk` is created. However if you find it more readable, the following is als possible:
```suggestion
    std::shared_lock read_lock(overcommit_m);
    if (cancellation_state == QueryCancellationState::NONE)
        return;
   
   read_lock.unlock();
    
    std::lock_guard lk(overcommit_m);
    if (cancellation_state != QueryCancellationState::NONE)
    {
        freed_memory += amount;
        if (freed_memory >= required_memory)
            releaseThreads();
    }
```

Reply:
Sure. It looks good to me. I'm trying to commit your suggestion but it tells `it can not apply suggestions on deleted lines`. Maybe we should remove that from line 121 to line 123.

Reply:
Hi @antaljanosbenjamin , I have pushed another commit according your suggestion. Please help to review it.